R/conditions.R

```diff
@@ -130,6 +130,63 @@ captureStackTraces <- function(expr) {
 #' @include globals.R
 .globals$deepStack <- NULL
 
+# (Warning, this is pretty crazy)
+#
+# appendWithLimit lets us append an element to a list, but if the list starts to
+# exceed a limit, it will start eliding elements from the middle of the list.
+# At the point of elision, it will insert a placeholder integer that indicates
+# how many elements were elided.
+#
+# This is useful for storing deep stack traces, where we don't want to overwhelm
+# the user with a huge number of stack traces, but both the head and tail may
+# contain important information.
+#
+# Example:
+# lst <- list("a", "b", "c", "d")
+#
+# appendWithLimit(lst, "e", limit = 4)
+# # Result: list("a", 1L, "c", "d", "e")
+#
+# appendWithLimit(lst, "e", limit = 4, keep_head = 2L)
+# # Result: list("a", "b", 1L, "d", "e")
+appendWithLimit <- function(lst, x, limit, keep_head = 1L) {
+  # Sanity check that the parameters make sense
+  stopifnot(keep_head > 0)
+  stopifnot(limit > keep_head)
+  stopifnot(!is.null(x))
+  stopifnot(is.null(lst) || is.list(lst))
+
+  elide_count_index <- which(vapply(lst, is.integer, logical(1)))
+  elide_count <- sum(as.integer(lst[elide_count_index]))
+
+  # Sanity check -- shouldn't have more than one elide placeholder
+  stopifnot(length(elide_count_index) <= 1L)
+  if (length(elide_count_index) == 1L) {
+    # For a given list, keep_head must remain constant between calls to
+    # appendWithLimit; that is, the elide count, if one exists, must be exactly
+    # where we expect it to be
+    stopifnot(elide_count_index == keep_head + 1L)
+  }
+
+  element_count <- length(lst) - length(elide_count_index)
+  if (element_count < limit) {
+    # No need to elide anything, just append and return
+    c(lst, list(x))
+  } else {
+    # The list may or may not already contain an elide count. If it does, we
+    # need to increment it. If it doesn't, we need to add one.
+
+    # Keep the first `keep_head` elements...
+    prefix <- lst[1:keep_head]
```

Comment:
For consistency with the `tail()` below, and also to avoid possibly `NULL` values in `prefix`

```suggestion
    prefix <- head(lst, keep_head)
```